Provide a way to customize `CoroutineContext` for gRPC-Kotlin (#4866)
line/armeria: site/src/pages/docs/advanced-kotlin.mdx

    suspend fun getUserTokenBlocking(@Param("name") name: String): String {
        val id = myReactiveClient.getId(name).await()
        val user = myBlockingRepository.findById(id)
        return user.token
    }
}
```

<Tip>

If you use [Logback integration](/docs/advanced-logging) or [Zipkin integration](/docs/advanced-zipkin),
coroutines should be executed in a thread that is aware of <type://RequestContext>.

</Tip>

### Customizing `CoroutineContext` for annotated services

Armeria provides a way to configure a coroutine context of annotated service methods
using <type://CoroutineContextService> decorator.

```kotlin
import com.linecorp.armeria.server.kotlin.CoroutineContextService

serverBuilder
    .annotatedService()
    .decorator(CoroutineContextService.newDecorator { ctx ->
        CoroutineName(ctx.config().defaultServiceNaming().serviceName(ctx) ?: "name")
    })
    .build(MyAnnotatedService())
```

## Coroutines support for [gRPC-Kotlin](https://github.com/grpc/grpc-kotlin)

<Tip>

Visit [grpc-kotlin](https://github.com/line/armeria/examples/grpc-kotlin) to check out a fully working example.

</Tip>

### Running a gRPC-Kotlin service

> gRPC kotlin can be used by registering coroutine stub services in [the same way](/docs/server-grpc) as using

Reviewer: ```suggestion
gRPC-Kotlin support is enabled automatically by registering coroutine stub services in [the same way](/docs/server-grpc) as using
```

    .serviceUnder("/docs", DocService())
    .build()
```

### Calling a gRPC-Kotlin service

You can use <type://GrpcClients> to specify a coroutine stub that ends with `CoroutineStub` when creating a 
gRPC-Kotlin client.

```kotlin
val client =
  GrpcClients
    .builder("https://grpc.service.com")
    .intercept(yourInterceptors)
    .build(YourServiceGrpcKt.YourServiceCoroutineStub::class.java)
```

<Tip>

Please refer to [Calling a gRPC service](/docs/client-grpc) for more information on gRPC clients.

</Tip>

### Customizing `CoroutineContext` for gRPC-Kotlin services

When you use gRPC-Kotlin's suspend functions, the default `CoroutineContext` is the combination of
<type://ServiceRequestContext> and an event loop `CoroutineDispatcher`. So you can access 
<type://ServiceRequestContext> from the suspend functions by using <type://ServiceRequestContext#current()>, 
similar to how you access it in annotated services.

With Armeria's integration with gRPC-Kotlin, it's easy to set up a custom `CoroutineContext` for your
gRPC-Kotlin service. You can create your own `CoroutineContext` using <type://CoroutineContextProvider>, and 
add it to your service through Java SPI.

> Note that the ServiceRequestContext is still part of the 

Reviewer: ```suggestion
Note that the <type://ServiceRequestContext> is still a part of the 
```

`CoroutineContext`, but the event loop `CoroutineDispatcher` is disabled if the custom `CoroutineContext`
is used.

```kotlin
class CustomCoroutineContextProvider : CoroutineContextProvider {
    override fun provide(ctx: ServiceRequestContext): CoroutineContext {
      return ... // A custom CoroutineContext
    }
}
```
<Tip>

> A custom `CoroutineContextProvider` should be registered as a Java SPI to
> `META-INF/services/com.linecorp.armeria.common.kotlin.CoroutineContextProvider`.

Reviewer: ```suggestion
A custom `CoroutineContextProvider` should be registered via Java SPI in
`META-INF/services/com.linecorp.armeria.common.kotlin.CoroutineContextProvider`.
```
